combination.go

```diff
@@ -0,0 +1,125 @@
+package errors
+
+import (
+	"fmt"
+	"io"
+	"strings"
+)
+
+// Combination combines multiple errors into one.
+// The Error method returns the strings from the individual Error methods
+// joined by the new line character '\n'.
+// Combination always has at least one error and returns the first error
+// as result of the Cause method.
+type Combination struct {
+	errs []error
+	*stack
+}
+
+// Combine returns a Combination error for 2 or more errors which are not nil,
+// or the callstack wrapped error if only one error was passed,
+// or nil if zero arguments are passed or all passed errors are nil.
+// Any returned non nil error will be wrapped with a callstack,
+// independent if a single error or a Combination type is returned.
+// The Combination type's Error method returns the strings from the
+// individual Error methods joined by the new line character '\n'.
+// Note that Cause(error) can only return a single error,
+// so in case of a Combination type, Cause returns the first error.
+func Combine(errs ...error) error {
+	switch len(errs) {
+	case 0:
+		return nil
+	case 1:
+		return WithStackSkip(1, errs[0])
+	}
+
+	var lastNotNil error
+	numNotNil := 0
+	for _, err := range errs {
+		if err != nil {
+			lastNotNil = err
+			numNotNil++
+		}
+	}
+	switch numNotNil {
+	case 0:
+		return nil
+	case 1:
+		return WithStackSkip(1, lastNotNil)
+	}
+
+	if numNotNil < len(errs) {
```

Comment:
`append` does a very good job of amortizing allocation costs. As such, it is strongly recommended that one simplify their code rather than attempt to "squeeze every ounce of performance" out of this.

This whole function would be much more readable as:
```
var collect []error
for _, err := range errs {
  if err != nil {
    collect = append(collect, err)
  }
}
switch len(collect) {
case 0;
  return nil
case 1:
  return WithStackSkip(1, collect[0])
}
return &combination{
  errs: errs,
  stack: callers(0),
}
```